### PR TITLE
Added idb 60 keyboard

### PR DESCRIPTION
**Pull request by itspngu**
## Description

Added idb 60 keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/idb_60/config.h [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONFIG_H
```
Comment (noroadsleft) [inline keyboards/idb_60/config.h]: If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`?

It's simpler, and less prone to user error, so if you could make this switch?

```suggestion
#pragma once
```

File: keyboards/idb_60/config.h [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+#define CONFIG_H
```
Comment (noroadsleft) [inline keyboards/idb_60/config.h]:
```suggestion
```

File: keyboards/idb_60/config.h [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+*/
+#define MATRIX_ROW_PINS { C2, D0, D1, D2, D3, D4, D5, D6, B0, B1 }
+#define MATRIX_COL_PINS { B2, B3, B4, C6, B6, B7, C7, B5 }
+#define UNUSED_PINS
+
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+/* Debounce reduces chatter (unintended double-presses) - set 0 if debouncing is not needed */
+#define DEBOUNCING_DELAY 5
```
Comment (noroadsleft) [inline keyboards/idb_60/config.h]: This definition's name has been changed to `DEBOUNCE`.

```suggestion
#define DEBOUNCE 5
```

File: keyboards/idb_60/config.h [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+ * the keyboard. They are best used in combination with the HID Listen program,
+ * found here: https://www.pjrc.com/teensy/hid_listen.html
+ *
+ * The options below allow the magic key functionality to be changed. This is
+ * useful if your keyboard/keypad is missing keys and you want magic key support.
+ *
+ */
+
+/* key combination for magic key command */
+#define IS_COMMAND() ( \
```
Comment (noroadsleft) [inline keyboards/idb_60/config.h]: Lines 100-102 should be commented out or deleted.

This definition is only necessary if you aren't using the default value.

https://github.com/qmk/qmk_firmware/blob/7209266ea1bd8b6bce4d468ea016689602499239/tmk_core/common/command.h#L37-L39

File: keyboards/idb_60/config.h [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+#define EEPROM_VERSION_ADDR 34
+
+// Dynamic keymap starts after EEPROM version
+#define DYNAMIC_KEYMAP_EEPROM_ADDR 35
+// Dynamic macro starts after dynamic keymaps (35+(4*10*6*2)) = (35+480)
+#define DYNAMIC_KEYMAP_MACRO_EEPROM_ADDR 675    // **** CHANGE THIS BASED ON MATRIX_ROWS & MATRIX_COLS ****
+#define DYNAMIC_KEYMAP_MACRO_EEPROM_SIZE 349    // **** CHANGE THIS BASED ON 1024-DYNAMIC_KEYMAP_MACRO_EEPROM_ADDR ****
+#define DYNAMIC_KEYMAP_MACRO_COUNT 16
+
+#endif
```
Comment (noroadsleft) [inline keyboards/idb_60/config.h]:
```suggestion
```

File: keyboards/idb_60/idb_60.h [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#ifndef _IDB_60_H
```
Comment (noroadsleft) [inline keyboards/idb_60/idb_60.h]: If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`?

It's simpler, and less prone to user error, so if you could make this switch?

```suggestion
#pragma once
```

File: keyboards/idb_60/idb_60.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+#define _IDB_60_H
```
Comment (noroadsleft) [inline keyboards/idb_60/idb_60.h]:
```suggestion
```

File: keyboards/idb_60/idb_60.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    { K40, K41, K42, K43, K44, K45, KC_NO }, \
+    { K50, K51, K52, K53, K54, K55, K56 }, \
+    { K60, K61, K62, K63, K64, K65, K66 }, \
+    { K70, K71, K72, K73, K74, K75, K76 }, \
+	{ K80, K81, KC_NO, KC_NO, KC_NO, KC_NO, K86 }, \
+    { K90, KC_NO, KC_NO, K93, KC_NO, K95, K96 } \
+}
+
+
+#endif
```
Comment (noroadsleft) [inline keyboards/idb_60/idb_60.h]:
```suggestion
```

File: keyboards/idb_60/idb_60.h [hunk 9]
```diff
@@ -0,0 +1,33 @@
+#ifndef _IDB_60_H
+#define _IDB_60_H
+
+#include "quantum.h"
+#include "led.h"
```
Comment (noroadsleft) [inline keyboards/idb_60/idb_60.h]: `quantum.h` includes `led.h` automatically.

```suggestion
```

File: keyboards/idb_60/idb_60.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+inline void _idb_60_caps_led_on(void)      { DDRC |=  (1<<5); PORTC &= ~(1<<5); }
+inline void _idb_60_esc_led_on(void)    	{ DDRC |=  (1<<4); PORTC &= ~(1<<4); }
+
+inline void _idb_60_caps_led_off(void)     { DDRC &= ~(1<<5); PORTC &= ~(1<<5); }
+inline void _idb_60_esc_led_off(void)   	{ DDRC &= ~(1<<4); PORTC &= ~(1<<4); }
+
+#define LAYOUT( \
```
Comment (noroadsleft) [inline keyboards/idb_60/idb_60.h]: The spacing on this layout macro is inconsistent. It'll work, but it doesn't look polished.

The electrical matrix (Lines 20-29) also only defines 7 positions for rows 1-9, even though `MATRIX_COLS` is 8. Obviously `K07` is the only position on that column, and the code still works this way.

I'd prefer if these issues were tidied up, but you don't absolutely have to.

File: keyboards/idb_60/idb_60.h [hunk 11]
```diff
@@ -0,0 +1,33 @@
+#ifndef _IDB_60_H
+#define _IDB_60_H
+
+#include "quantum.h"
+#include "led.h"
+
+inline void _idb_60_caps_led_on(void)      { DDRC |=  (1<<5); PORTC &= ~(1<<5); }
```
Comment (noroadsleft) [inline keyboards/idb_60/idb_60.h]: You should be able to use the [GPIO Commands](https://docs.qmk.fm/#/internals_gpio_control) for Lines 7-11:

```c
inline void _idb_60_caps_led_on(void) {
    setPinOutput(C5);
    writePinLow(C5);
}

inline void _idb_60_esc_led_on(void) {
    setPinOutput(C4);
    writePinLow(C4);
}

inline void _idb_60_caps_led_off(void) {
    setPinInput(C5);
    writePinLow(C5);
}

inline void _idb_60_esc_led_off(void) {
    setPinInput(C4);
    writePinLow(C4);
}
```

File: keyboards/idb_60/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,35 @@
+#include QMK_KEYBOARD_H
+
+#define _______ KC_TRNS
```
Comment (noroadsleft) [inline keyboards/idb_60/keymaps/default/keymap.c]: QMK defines this keycode alias automatically. Line 3 can be deleted.

```suggestion
```

File: keyboards/idb_60/keymaps/default/keymap.c [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+	_______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+	_______, _______, _______,                                     _______,                            _______, _______, _______
+  )
+
+};
+
+void matrix_scan_user(void) {
+
+//Layer LED indicators
+    uint32_t layer = layer_state;
```
Comment (noroadsleft) [inline keyboards/idb_60/keymaps/default/keymap.c]: `layer_state` isn't always a `uint32_t`; we have a new typedef for this that automatically matches whatever's necessary for the keymap.

```suggestion
    layer_state_t layer = layer_state;
```

File: keyboards/idb_60/keymaps/pngu/keymap.c [hunk 14]
```diff
@@ -0,0 +1,42 @@
+#include QMK_KEYBOARD_H
+
+#define _______ KC_TRNS
```
Comment (noroadsleft) [inline keyboards/idb_60/keymaps/pngu/keymap.c]:
```suggestion
```

File: keyboards/idb_60/keymaps/pngu/keymap.c [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+	_______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+	_______, _______, _______,                                     _______,                            _______, _______, _______
+  )
+
+};
+
+void matrix_scan_user(void) {
+
+//Layer LED indicators
+    uint32_t layer = layer_state;
```
Comment (noroadsleft) [inline keyboards/idb_60/keymaps/pngu/keymap.c]:
```suggestion
    layer_state_t layer = layer_state;
```

File: keyboards/idb_60/readme.md [hunk 16]
```diff
@@ -0,0 +1,17 @@
+# idb 60 PCB
+
+Firmware for the idb 60 PCB
+
+Keyboard Maintainer: [/u/omgitspngu](https://github.com/itspngu)
```
Comment (noroadsleft) [inline keyboards/idb_60/readme.md]: Lines 5 and 6 need to end in two spaces for GitHub to render them correctly.

```suggestion
Keyboard Maintainer: [/u/omgitspngu](https://github.com/itspngu)  
```

File: keyboards/idb_60/readme.md [hunk 17]
```diff
@@ -0,0 +1,17 @@
+# idb 60 PCB
+
+Firmware for the idb 60 PCB
+
+Keyboard Maintainer: [/u/omgitspngu](https://github.com/itspngu) 
+Hardware Supported: idb 60 PCB
```
Comment (noroadsleft) [inline keyboards/idb_60/readme.md]:
```suggestion
Hardware Supported: idb 60 PCB  
```

File: keyboards/idb_60/rules.mk [hunk 18]
```diff
@@ -0,0 +1,66 @@
+# MCU name
+#MCU = at90usb1287
```
Comment (noroadsleft) [inline keyboards/idb_60/rules.mk]:
```suggestion
```

File: keyboards/idb_60/rules.mk [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+#
+#     If no clock division is performed on the input clock inside the AVR (via the
+#     CPU clock adjust registers or the clock division fuses), this will be equal to F_CPU.
+F_USB = $(F_CPU)
+
+# Interrupt driven control endpoint task(+60)
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+
+# Boot Section Size in *bytes*
```
Comment (noroadsleft) [inline keyboards/idb_60/rules.mk]: If you know the bootloader that's being used, it should be defined here.

https://github.com/qmk/qmk_firmware/blob/6c7c178cd2b9d535e76a0cd9dd3d1293b8ba1ed2/quantum/template/avr/rules.mk#L42-L59

File: keyboards/idb_60/idb_60.c [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+extern inline void _idb_60_caps_led_on(void);
+extern inline void _idb_60_esc_led_on(void);
+
+extern inline void _idb_60_caps_led_off(void);
+extern inline void _idb_60_esc_led_off(void);
+
+
+void led_set_kb(uint8_t usb_led) {
+
+    if (usb_led & (1<<USB_LED_CAPS_LOCK)) {
```
Comment (fauxpark) [inline keyboards/idb_60/idb_60.c]:
```suggestion
    if (IS_LED_ON(usb_led, USB_LED_CAPS_LOCK)) {
```

File: keyboards/idb_60/keymaps/default/keymap.c [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+
+    if (layer & (1<<1)) {
```
Comment (fauxpark) [inline keyboards/idb_60/keymaps/default/keymap.c]:
```suggestion
    if (layer_state_is(1)) {
```

No need to grab the layer state above, I think.

File: keyboards/idb_60/keymaps/pngu/keymap.c [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+    if ((layer & (1<<1)) || (layer & (1<<2))) {
```
Comment (fauxpark) [inline keyboards/idb_60/keymaps/pngu/keymap.c]:
```suggestion
    if (layer_state_is(1) || layer_state_is(2)) {
```

Same here.

**Comment by itspngu** [comment]
Thanks for the help guys. Implemented your suggestions.

File: keyboards/idb_60/config.h [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID    0x6060 // 24672
+#define PRODUCT_ID   0x6060 // 24672
+#define DEVICE_VER   0x0001
+#define MANUFACTURER    pngu
+#define PRODUCT         idb 60
+#define DESCRIPTION     t.m.k. keyboard firmware for idb 60
```
Comment (noroadsleft) [inline keyboards/idb_60/config.h]:
```suggestion
#define DESCRIPTION     QMK keyboard firmware for idb 60
```

File: keyboards/idb_60/rules.mk [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+#   Teensy halfKay   512
+#   Teensy++ halfKay 1024
+#   Atmel DFU loader 4096
+#   LUFA bootloader  4096
+#   USBaspLoader     2048
+BOOTLOADER = atmel-dfu
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
```
Comment (noroadsleft) [inline keyboards/idb_60/rules.mk]: Line 48 can be deleted. The `BOOTLOADER` rule on Line 47 will set this automatically, so it's redundant here.

```suggestion
```

**Comment by itspngu** [comment]
Thanks for bearing with me, this is my first PR. I think I fixed everything now?

**Comment by noroadsleft** [comment]
Spacing is still weird on the `idb_60.h` file, but I'm going to let that go.

**Comment by itspngu** [comment]
Is everything okay now? I'm not sure what the process is like. Can this be merged? :)

Have a nice weekend y'all.

File: keyboards/idb_60/idb_60.h [hunk 25]
```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include "quantum.h"
+
+inline void _idb_60_caps_led_on(void) {
+    setPinOutput(C5);
```
Comment (drashna) [inline keyboards/idb_60/idb_60.h]: You do not, and should not set the output with each on/off.  It only needs to be set on initialization.

```suggestion
```

File: keyboards/idb_60/idb_60.h [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+    writePinLow(C5);
+}
+
+inline void _idb_60_esc_led_on(void) {
+    setPinOutput(C4);
```
Comment (drashna) [inline keyboards/idb_60/idb_60.h]:
```suggestion
```

File: keyboards/idb_60/idb_60.h [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+    writePinLow(C4);
+}
+
+inline void _idb_60_caps_led_off(void) {
+    setPinInput(C5);
```
Comment (drashna) [inline keyboards/idb_60/idb_60.h]:
```suggestion
```

File: keyboards/idb_60/idb_60.h [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+    writePinLow(C5);
+}
+
+inline void _idb_60_esc_led_off(void) {
+    setPinInput(C4);
```
Comment (drashna) [inline keyboards/idb_60/idb_60.h]:
```suggestion
```

File: keyboards/idb_60/idb_60.h [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+    writePinLow(C4);
+}
+
```
Comment (drashna) [inline keyboards/idb_60/idb_60.h]: But you'll need to initialize the pins: 

```suggestion

void keyboarb_pre_init_kb(void) {
    setPinOutput(C4);
    setPinOutput(C5);
}
```

File: keyboards/idb_60/keymaps/default/keymap.c [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+	RESET,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,  _______, _______,
+	_______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+	_______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
+	_______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+	_______, _______, _______,                                     _______,                            _______, _______, _______
+  )
+
+};
+
+void matrix_scan_user(void) {
```
Comment (drashna) [inline keyboards/idb_60/keymaps/default/keymap.c]: This actually could be done with:

```c
layer_state_t layer_state_set_user(layer_state_t state) {
    if (layer_state_cmp(state, 1) {
        _idb_60_esc_led_on();
    } else {
        _idb_60_esc_led_off();
    }
}
```

File: keyboards/idb_60/idb_60.h (hunk 29 above)
Comment (noroadsleft) [reply keyboards/idb_60/idb_60.h]: `keyboard_pre_init_kb` *

File: keyboards/idb_60/idb_60.h [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+inline void _idb_60_caps_led_off(void) {
+    writePinLow(C5);
+}
+
+inline void _idb_60_esc_led_off(void) {
+    writePinLow(C4);
+}
+
+inline void keyboarb_pre_init_kb(void) {
```
Comment (noroadsleft) [inline keyboards/idb_60/idb_60.h]: spell check

```suggestion
inline void keyboard_pre_init_kb(void) {
```

Comment (drashna) [reply keyboards/idb_60/idb_60.h]: Whoops, sorry.

Comment (drashna) [reply keyboards/idb_60/idb_60.h]: Also, this shouldn't be "inline"

Comment (itspngu) [reply keyboards/idb_60/idb_60.h]: Oh, that slipped past me. Fixed.

Comment (drashna) [reply keyboards/idb_60/idb_60.h]: Yeah, slipped past me too, until the second check.

**Comment by itspngu** [comment]
@noroadsleft @drashna @fauxpark Hey! I'm not sure how this works or why it's still saying 'changes requested' in red, but as far as I can tell I did everything you guys suggested. Would you mind giving this another look so it can be merged? Have a nice weekend, cheers, -pngu

**Pull request merged**